Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,416 @@
+
+@[simp]
+theorem liftOn_mk {α : Type*} (m : M) (s : ℕ+)
+    (f : M → ℕ+ → α)
+    (h : ∀ (m m' : M) (s s' : ℕ+), mk m s = mk m' s' → f m s = f m' s') :
+    liftOn (mk m s) f h = f m s := rfl
+
+/-- If `f : M → ℕ+ → M → ℕ+ → α` respects the equivalence on
+localization, lift it to a function `DivisibleHull M → DivisibleHull M → α`. -/
+def liftOn₂ {α : Type*} (x y : DivisibleHull M)
+    (f : M → ℕ+ → M → ℕ+ → α)
+    (h : ∀ (m n m' n' : M) (s t s' t' : ℕ+),
+      mk m s = mk m' s' → mk n t = mk n' t' → f m s n t = f m' s' n' t') : α :=
+  LocalizedModule.liftOn₂ x y (fun p q ↦ f p.1 (↑ⁿ.symm p.2) q.1 (↑ⁿ.symm q.2))
+    fun p q p' q' heq heq' ↦
+    h p.1 q.1 p'.1 q'.1 (↑ⁿ.symm p.2) (↑ⁿ.symm q.2) (↑ⁿ.symm p'.2) (↑ⁿ.symm q'.2)
+      (by
+        obtain ⟨u, hu⟩ := heq
+        exact mk_eq_mk.mpr ⟨↑ⁿ.symm u, hu⟩)
+      (by
+        obtain ⟨u, hu⟩ := heq'
+        exact mk_eq_mk.mpr ⟨↑ⁿ.symm u, hu⟩)
+
+@[simp]
+theorem liftOn₂_mk {α : Type*} (m m' : M) (s s' : ℕ+)
+    (f : M → ℕ+ → M → ℕ+ → α)
+    (h : ∀ (m n m' n' : M) (s t s' t' : ℕ+),
+      mk m s = mk m' s' → mk n t = mk n' t' → f m s n t = f m' s' n' t') :
+    liftOn₂ (mk m s) (mk m' s') f h = f m s m' s' := rfl
+
+theorem mk_add_mk {m1 m2 : M} {s1 s2 : ℕ+} :
+    mk m1 s1 + mk m2 s2 = mk (s2.val • m1 + s1.val • m2) (s1 * s2) := LocalizedModule.mk_add_mk
+
+theorem mk_add_mk_left {m1 m2 : M} {s : ℕ+} :
+    mk m1 s + mk m2 s = mk (m1 + m2) s := by
+  rw [mk_add_mk, mk_eq_mk]
+  exact ⟨1, by simp [smul_smul]⟩
+
+@[simp]
+theorem coe_add {m1 m2 : M} : ↑(m1 + m2) = (↑m1 + ↑m2 : DivisibleHull M) := by simp [mk_add_mk_left]
+
+variable (M) in
+/-- Coercion from `M` to `DivisibleHull M` as a `AddMonoidHom`. -/
+@[simps]
+def coeAddMonoidHom : M →+ DivisibleHull M where
+  toFun := (↑)
+  map_zero' := by simp
+  map_add' := by simp
+
+theorem nsmul_mk (a : ℕ) (m : M) (s : ℕ+) : a • mk m s = mk (a • m) s := by
+  induction a with
+  | zero => simp
+  | succ n h => simp [add_nsmul, mk_add_mk_left, h]
+
+theorem nnqsmul_mk (a : ℚ≥0) (m : M) (s : ℕ+) :
+    a • mk m s = mk (a.num • m) (⟨a.den, a.den_pos⟩ * s) := by
+  convert LocalizedModule.mk'_smul_mk ℚ≥0 a.num m ⟨a.den, by simp⟩ (↑ⁿ s)
+  simp [IsLocalization.eq_mk'_iff_mul_eq]
+
+section TorsionFree
+variable [IsAddTorsionFree M]
+
+theorem mk_eq_mk_iff_smul_eq_smul {m m' : M} {s s' : ℕ+} :
+    mk m s = mk m' s' ↔ s'.val • m = s.val • m' := by
+  aesop (add simp [mk_eq_mk, nsmul_right_inj])
+
+theorem mk_left_injective (s : ℕ+) : Function.Injective (fun (m : M) ↦ mk m s) := by
+  intro m n h
+  simp_rw [mk_eq_mk_iff_smul_eq_smul] at h
+  exact nsmul_right_injective (by simp) h
+
+theorem coe_injective : Function.Injective ((↑) : M → DivisibleHull M) :=
+  mk_left_injective 1
+
+theorem coeAddMonoidHom_injective : Function.Injective (coeAddMonoidHom M) := coe_injective
+
+end TorsionFree
+
+section Group
+variable {M : Type*} [AddCommGroup M]
+
+@[simp]
+theorem mk_neg (m : M) (s : ℕ+) : mk (-m) s = -mk m s :=
```

Comment:
Given the direction of `nsmul_mk` and `zsmul_mk`, shouldn't this be turned around?